### PR TITLE
Add js extension to preset-env generated polyfill imports

### LINKED ISSUE #12378: core-js 3 missing .js extension when 'type': 'module'
## Bug Report

- [ ] I would like to work on a fix!

**Current behavior**

`'type': 'module'` is used in `package.json` and when `corejs: 3` is specified in `babel.config.js` webpack complains about missing `.js` extension despite it existing in `src/page.js`. If I comment out `corejs: 3` then the error goes away.

```
ERROR in ./src/page.js 1:0-37
Module not found: Error: Can't resolve 'core-js/modules/es.weak-set' in '/home/kieran/PhpstormProjects/babel-test/src'
Did you mean 'es.weak-set.js'?
BREAKING CHANGE: The request 'core-js/modules/es.weak-set' failed to resolve only because it was resolved as fully specified
(probably because the origin is a '*.mjs' file or a '*.js' file where the package.json contains '"type": "module"').
The extension in the request is mandatory for it to be fully specified.
Add the extension to the request.
```

I would be incredibly grateful if someone could put me out my misery.

**Input Code**

https://github.com/bytestream/babel-test

**Expected behavior**

I expect the code to transpile when using corejs 3.

**Babel Configuration (babel.config.js, .babelrc, package.json#babel, cli command, .eslintrc)**
- Filename: `babel.config.js`

```js
module.exports = {
    presets: [
        [
            '@babel/preset-env',
            {
                'useBuiltIns': 'entry',
                'corejs': 3, // <--- problematic line
            }
        ]
    ],
    plugins: [
        // Don't allow polyfills to pollute global namespace.
        '@babel/plugin-transform-runtime',
        // Support for the experimental syntax 'classPrivateProperties'
        '@babel/plugin-proposal-class-properties',
        // Support for the experimental syntax 'classPrivateMethods'
        '@babel/plugin-proposal-private-methods',
    ]
};
```

**Environment**

```
$ npx envinfo --preset babel

  System:
    OS: Linux 4.9 Debian GNU/Linux 9 (stretch) 9 (stretch)
  Binaries:
    Node: 12.19.0 - /usr/bin/node
    Yarn: 1.22.4 - /usr/bin/yarn
    npm: 6.14.8 - /usr/bin/npm
  npmPackages:
    @babel/core: ^7.12.3 => 7.12.3 
    @babel/plugin-proposal-class-properties: ^7.12.1 => 7.12.1 
    @babel/plugin-proposal-private-methods: ^7.12.1 => 7.12.1 
    @babel/plugin-transform-runtime: ^7.12.1 => 7.12.1 
    @babel/preset-env: ^7.12.1 => 7.12.1 
    @babel/runtime-corejs3: ^7.12.5 => 7.12.5 
    babel-loader: ^8.2.1 => 8.2.1 
    webpack: ^5.6.0 => 5.6.0 
```

**Possible Solution**

**Additional context**

### PR DESCRIPTION
**Pull request by JLHwung**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | `preset-env` generated esm polyfill imports will throw on Node.js 13, fixes #12378 
| Patch: Bug Fix?          | Yes
| Tests Added + Pass?      | Yes
| License                  | MIT

In this PR we add `.js` extension to the injected polyfill imports by `@babel/preset-env`. The new import specifier `path/to/polyfill.js` could cause trouble for AMD module compatibility (see #10832 ) but practically it is okay in this scenario because both our upstream polyfills: core-js and regenerator do not support amd.

Note that this PR is different than #10549 which handles `transform-runtime` instead.

/cc @rwjblue 

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/10862"></a>

### CODE REVIEW

File: packages/babel-preset-env/src/utils.js [hunk 1]
```diff
@@ -152,11 +152,11 @@ export function isPolyfillSource(source: ?string): boolean {
 }
 
 const modulePathMap = {
-  "regenerator-runtime": "regenerator-runtime/runtime",
+  "regenerator-runtime": "regenerator-runtime/runtime.js",
```
Comment (nicolo-ribaudo) [inline packages/babel-preset-env/src/utils.js:155]: `regenerator-runtime` works in browsers, so it currently works with AMD I think.

Comment (JLHwung) [reply packages/babel-preset-env/src/utils.js:60]: `regenerator-runtime/runtime.js` does not use `define` to declare its name and dependencies, if any. So I guess it is likely that no one consume it as an AMD module.

**Comment by nicolo-ribaudo** [comment]
You need to patch the vue e2e test similarly to what we did in `e2e-create-react-app.sh`, since this PR is making those tests failing.

@sodatea This PR will make `vue-cli` tests fail (https://app.circleci.com/pipelines/github/babel/babel/4604/workflows/bae839be-16f9-4ea1-975b-2477fb57eaff/jobs/33616/parallel-runs/0/steps/0-103), since they test that the code produced by Babel contains an exact substring changed by this PR.

**Comment by sodatea** [comment]
@nicolo-ribaudo Thanks for the caution! I've fixed the tests.

**Review by nicolo-ribaudo — APPROVED** [review]

**Comment by existentialism** [comment]
@sodatea mind fixing the conflicts?

**Comment by nicolo-ribaudo** [comment]
@existentialism I think you pinged the wrong person :stuck_out_tongue: 

Btw, fixing the conflicts probably also fixes the e2e tests, since they are failing exactly in those two files.

**Comment by nicolo-ribaudo** [comment]
Spammish, but tomorrow it's the birthday of this PR :partying_face:

**Review by existentialism — APPROVED** [review]

**Pull request merged**